File: src/main/java/com/boschchicagoconnectory/takehome/joan1stProblem/Joan1stProblem.java
package com.boschchicagoconnectory.takehome.joan1stProblem;

import com.boschchicagoconnectory.takehome.joan1stProblem.util.JoinArrays;
import org.json.*;

import java.io.*;
import java.util.*;

public class Joan1stProblem {
    public static void main(String[] args) throws IOException {
        Map<Integer, Object> jsonArray = new HashMap<>();

        for(String file: args) {
            FileInputStream fis = new FileInputStream(file);

            if(file.toLowerCase().endsWith(".json")) {
                JSONArray jArr = new JSONArray(new JSONTokener(fis));
                new JoinArrays(jsonArray, jArr).join();

            }else if(file.toLowerCase().endsWith(".csv")){
                JSONArray jArr = CDL.toJSONArray(new JSONTokener(fis));
                new JoinArrays(jsonArray, jArr).join();

            }else if(file.toLowerCase().endsWith(".txt")){
                BufferedReader bis = new BufferedReader(new InputStreamReader(fis));
                String line = bis.readLine();
                StringBuilder sb = new StringBuilder();
                while (line != null){
                    String newLine = "\""+line.replaceAll("\t", "\",\"")+"\"\n";
                    sb.append(newLine);
                    line = bis.readLine();
                }
                JSONArray jArr = CDL.toJSONArray(sb.toString());
                new JoinArrays(jsonArray, jArr).join();
            }else{
                throw new FileNotFoundException("The file args should have the following extensions: '.json', '.csv', '.txt'");
            }

            fis.close();
        }
        List jsonArrayList = new ArrayList<>(jsonArray.values());
        Collections.sort(jsonArrayList, new Comparator<JSONObject>() {

            @Override
            public int compare(JSONObject jsonObjectA, JSONObject jsonObjectB) {
                int compare;
                String genreA = jsonObjectA.getString("Genre");
                String genreB = jsonObjectB.getString("Genre");
                compare = genreA.compareTo(genreB);
                if(compare == 0){
                    String authorA = jsonObjectA.getString("Author").split(" ")[1];
                    String authorB = jsonObjectB.getString("Author").split(" ")[1];
                    compare = authorA.compareTo(authorB);
                }
                if(compare == 0){
                    String authorA = jsonObjectA.getString("Author").split(" ")[0];
                    String authorB = jsonObjectB.getString("Author").split(" ")[0];
                    compare = authorA.compareTo(authorB);
                }

                return compare;
            }
        });

        String csv = CDL.toString(new JSONArray(jsonArrayList));
        FileWriter output = new FileWriter("FinalCSV.csv");
        output.write(csv);
        output.close();

    }
}

File: src/Joan1stProblem.kt
import org.json.CDL
import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener
import util.JoinArrays
import java.io.*
import java.util.*

class Joan1stProblem {
    fun main(args: Array<String>){
        val jsonArray: HashMap<Int, Any> = HashMap()

        for (file in args) {
            val fis = FileInputStream(file)
            if (file.toLowerCase().endsWith(".json")) {
                val jArr = JSONArray(JSONTokener(fis))
                JoinArrays(jsonArray, jArr).join()
            } else if (file.toLowerCase().endsWith(".csv")) {
                val jArr: JSONArray = CDL.toJSONArray(JSONTokener(fis))
                JoinArrays(jsonArray, jArr).join()
            } else if (file.toLowerCase().endsWith(".txt")) {
                val bis = BufferedReader(InputStreamReader(fis))
                var line = bis.readLine()
                val sb = StringBuilder()
                while (line != null) {
                    val newLine = "\"" + line.replace("\t".toRegex(), "\",\"") + "\"\n"
                    sb.append(newLine)
                    line = bis.readLine()
                }
                val jArr: JSONArray = CDL.toJSONArray(sb.toString())
                JoinArrays(jsonArray, jArr).join()
            } else {
                throw FileNotFoundException("The file args should have the following extensions: '.json', '.csv', '.txt'")
            }
            fis.close()
        }
        val jsonArrayList: List<JSONObject> = jsonArray.values as List<JSONObject>
        Collections.sort(jsonArrayList, Comparator<JSONObject> { jsonObjectA, jsonObjectB ->
            var compare: Int
            val genreA: String = jsonObjectA.getString("Genre")
            val genreB: String = jsonObjectB.getString("Genre")
            compare = genreA.compareTo(genreB)
            if (compare == 0) {
                val authorA: String = jsonObjectA.getString("Author").split(" ").get(1)
                val authorB: String = jsonObjectB.getString("Author").split(" ").get(1)
                compare = authorA.compareTo(authorB)
            }
            if (compare == 0) {
                val authorA: String = jsonObjectA.getString("Author").split(" ").get(0)
                val authorB: String = jsonObjectB.getString("Author").split(" ").get(0)
                compare = authorA.compareTo(authorB)
            }
            compare
        } as Comparator<in Any?>?)

        val csv: String = CDL.toString(JSONArray(jsonArrayList))
        val output = FileWriter("FinalCSV.csv")
        output.write(csv)
        output.close()


    }
}